wi-fi/hal/cyhal_sdio.c

```diff
@@ -0,0 +1,783 @@
+			return -1;
+		}
+
+		if ((val & flags) == flags)
+			return 0;
+
+		usleep(wait_us);
+	}
+
+	cy_log_msg(CYLF_SDIO, CY_LOG_WARNING, "SDIO cmd timeout\n");
+
+	return -1;
+}
+
+
+static int reset_all(uint8_t sdclkfs, uint8_t dvs)
+{
+	uint32_t i, val;
+
+	cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "reset SDIO\n");
+
+	*(sdio_common.base + sys_ctrl) = (1 << 24) | (sdclkfs << 8) | (dvs << 4) | 0xf; /* RSTA=1 */
+
+	// FIXME: ???
+	usleep(10);
+
+	for (i = 0; i < 10; ++i) {
+		val = *(sdio_common.base + sys_ctrl) & (1 << 24);
+		if (val == 0)
+			return 0;
+		usleep(10);
+	}
+
+	cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "failed to reset SDIO\n");
+
+	return -1;
+}
+
+
+static int reset_cmd_block(void)
+{
+	uint32_t i, val;
+
+	cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "reset SDIO cmd block\n");
+
+	val = *(sdio_common.base + sys_ctrl);
+	val |= 1 << 25; /* RSTC=1 */
+	*(sdio_common.base + sys_ctrl) = val;
+
+	// FIXME: ???
+	usleep(10);
+
+	for (i = 0; i < 10; ++i) {
+		val = *(sdio_common.base + sys_ctrl) & (1 << 25);
+		if (val == 0)
+			return 0;
+		usleep(10);
+	}
+
+	cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "failed to reset SDIO cmd block\n");
+
+	return -1;
+}
+
+
+static int reset_data_block(void)
+{
+	uint32_t i, val;
+
+	cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "reset SDIO data block\n");
+
+	val = *(sdio_common.base + sys_ctrl);
+	val |= 1 << 26; /* RSTD=1 */
+	*(sdio_common.base + sys_ctrl) = val;
+
+	// FIXME: ???
+	usleep(10);
+
+	for (i = 0; i < 10; ++i) {
+		val = *(sdio_common.base + sys_ctrl) & (1 << 26);
+		if (val == 0)
+			return 0;
+		usleep(10);
+	}
+
+	cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "failed to reset SDIO data block\n");
+
+	return -1;
+}
+
+
+static void sdio_irq_thread(void *arg)
+{
+	bool finish;
+	uint32_t val;
+
+	while (1) {
+		mutexLock(sdio_common.irq_lock);
+
+		while (1) {
+			finish = sdio_common.irq_thread_finish;
+			if (finish)
+				break;
+
+			val = *(sdio_common.base + int_status);
+			if (val & (1 << 8))
+				break;
+
+			condWait(sdio_common.irq_cond, sdio_common.irq_lock, 0);
+		}
+
+		mutexUnlock(sdio_common.irq_lock);
+
+		if (finish)
+			break;
+
+		// cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "got SDIO IRQ\n");
+
+		/* this will also clear CINT */
+		val = *(sdio_common.base + int_status_en);
+		val &= ~(1 << 8); /* CINTESEN=0 */
+		*(sdio_common.base + int_status_en) = val;
+
+		sdio_common.irq_handler(sdio_common.irq_handler_arg, CYHAL_SDIO_CARD_INTERRUPT);
+
+		val = *(sdio_common.base + int_status_en);
+		val |= 1 << 8; /* CINTESEN=1 */
+		*(sdio_common.base + int_status_en) = val;
+
+		val = *(sdio_common.base + int_signal_en);
+		val |= 1 << 8; /* CINTIEN=1 */
+		*(sdio_common.base + int_signal_en) = val;
+
+		// cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "done SDIO IRQ\n");
+	}
+
+	/* signal that we have finished */
+	mutexLock(sdio_common.irq_lock);
+	sdio_common.irq_thread_finish = false;
+	condSignal(sdio_common.irq_cond);
+	mutexUnlock(sdio_common.irq_lock);
+
+	endthread();
```

Comment:
thread join might be better here (so no need for condSignal, we need to do thread join anyway to clean it's resources)

Reply:
Could you point me to some example of thread join? I thought it was not implemented in RTOS.

Reply:
Pthreads implementation should be a nice example - in most cases we use worker threads not task threads, so threads never end.

https://github.com/phoenix-rtos/libphoenix/blob/c6660555f207fdfee573328190243819ec12b112/pthread/pthread.c#L127